### ISSUE TITLE
Multiple orders can't be tracked properly when using different carriers

### ISSUE THREAD

**Issue by PeeyushAgrawal** [PrestaShop/PrestaShop#22836]
#### Describe the bug

There are two product which has different shipping methods now if a guest place the order with these two products then guest is not able to follow both orders (orderreference#1  and orderreference#2) through the guest tracking page always getting open orderreference#1

#### Expected behavior

If the guest enter orderreference#1 then this order (# 1) details should visible and if the guest enter orderreference#2 then the order details should visible according  to orderreference#2

#### Steps to Reproduce

1. Set two standard product with different shipping methods
2. add these two products to the cart
3. Now complete the order as a guest
4. After completing you will get emails for different orders with the same reference
5. Now click on guest tracking and enter here order reference with # 1 or # 2 always you will get the order details that belongs to # 2, you will not able to check the order details for # 1 

**Screenshots**
Let's have an example
We have ordered "test product 1" and "test product 2" with the different shipping method as a guest and got two emails
![order1](https://user-images.githubusercontent.com/39614089/104620866-78c24900-56b5-11eb-853f-68429073382a.png)

and 
![order2](https://user-images.githubusercontent.com/39614089/104620897-7f50c080-56b5-11eb-830d-90c94721ed52.png)

Now when we enter VKEMMGUQG#1 in order reference then get the order details with test product 2 

![order1product](https://user-images.githubusercontent.com/39614089/104621362-04d47080-56b6-11eb-9666-14478813cf59.png)

And when we enter VKEMMGUQG#2 

![guesttracking](https://user-images.githubusercontent.com/39614089/104621528-38af9600-56b6-11eb-9db9-d5e3b28568b7.png)

then still getting the same order details with test product 2 instead os test product 1
#### Additional information

* PrestaShop version: 1.7.7.1
* PHP version: 7.3

**Comment by florine2623** [comment]
Hello @PeeyushAgrawalWebkul ,

Thanks for the detailed issue.
I was able to reproduce the issue with PrestaShop version 1.7.7.1 and PS 1.7.6.5. I’ll add this to the debug backlog so that it’s fixed.
If you have already fixed it on your end or if you think you can do it, please do send us a pull request!

Thanks!

**Comment by PeeyushAgrawal** [comment]
@florine2623 Yes, i have fixed it on my end and will send a pull request soon.

**Comment by boo-code** [comment]
Still reproduces on 9.2.x. @PeeyushAgrawal's #24195 was closed for inactivity rather than for anything wrong with the approach, so I have prepared a fix building on it.

`PaymentModule::validateOrder()` generates the reference once, before the package loop, and re-rolls until it is unused, so a reference identifies a **cart**, not an order. `Order::getUniqReference()` is what distinguishes the split orders, and it is what the emails print.

Guest tracking discards it on the first line of `postProcess()`:

```php
$order_reference = current(explode('#', Tools::getValue('order_reference')));
```

and the lookup then takes whichever row comes first, with no ORDER BY. The docblock on `getByReferenceAndEmail()` states the assumption that breaks here: "The combination (reference, email) should be unique, of multiple entries are found, then we take the first one."

One consequence that is not in the report: since the suffix is dropped entirely, **`REFERENCE#99` currently opens a real order**. Any number is accepted and ignored.

The detail that shapes the fix is that the suffix is not an ordinal:

```php
return $this->reference . '#' . ($this->id + 1 - $order['min']);
```

It is an offset from the lowest order id of **that order's cart**, so if another shopper checks out between two packages the ids are not contiguous and the numbers skip — two orders can be `#1` and `#3`. Selecting "the Nth row" would be wrong exactly when concurrent checkouts happen.

Rather than reimplement that arithmetic, the lookup asks `getUniqReference()` itself which order carries the number. The two group by different columns — the generator takes `MIN(id_order)` over `id_cart`, while a lookup can only start from the reference the customer typed — and those coincide only while one reference belongs to one cart. In core they do, but `$order_reference` is a public parameter of `validateOrder()` that a module could pass to reuse one. None of the callers in core or the bundled payment modules passes it, and delegating to the generator keeps the fix correct even if one did.

`getByReferenceAndEmail()` takes an optional third argument, keeping the signature compatible; guest tracking is its only caller in core.

Covered by a round-trip test: whatever `getUniqReference()` prints for an order has to resolve back to that order. One case forces the non-contiguous ids and asserts the suffix really is not `#2` before checking it still resolves. Reverting both files fails 3 of the 4 tests, including `34 is identical to 0` for the unknown suffix.

@PeeyushAgrawal, if you would rather pick your PR back up, say so and I will drop mine — it is your finding.

I am at the open-PR limit, so I cannot open it yet. I will link it here as soon as a slot frees.
